cli/fossilize_replay.cpp

```diff
@@ -2822,27 +2893,45 @@ struct ThreadedReplayer : StateCreatorInterface
 			                 }});
 
 			work.push_back({ get_order_index(ENQUEUE_OUT_OF_RANGE_PARENT_PIPELINES),
-			                 [this, &pipelines, derived]() {
+			                 [this, &pipelines, derived, parsed_parents]() {
 				                 // Figure out which of the parent pipelines we need.
 				                 for (auto &d : *derived)
-					                 enqueue_parent_pipelines(d, pipelines);
+					                 enqueue_parent_pipelines(d, pipelines, *parsed_parents);
 			                 }});
 
 			if (memory_index == 0)
 			{
 				// This is a join-like operation. We need to wait for all parent pipelines and all shader modules to have completed.
 				work.push_back({get_order_index(ENQUEUE_SHADER_MODULE_SECONDARY_OFFSET),
-				                [this, &parents, hash_offset, start_index]()
+				                [this, &pipelines, derived, parsed_parents, &hash_to_info, &parents, hash_offset, start_index]()
 				                {
-					                // Wait until all parent pipelines have been parsed.
+					                bool progress = true;
+					                while (progress)
+					                {
+						                progress = false;
+
+						                // Wait until all parent pipelines have been parsed.
+						                sync_worker_memory_context(PARENT_PIPELINE_MEMORY_CONTEXT);
+
+						                // Queue up all shader modules if somehow the shader modules used by parent pipelines differ from children ...
+						                for (auto &parent : parents)
+						                {
+							                // Handle nested libraries.
+							                progress |= enqueue_parent_pipelines(parent.second, pipelines, *parsed_parents);
```

Comment:
This is a race condition since enqueue_parent_pipelines will push work that ends up writing to `parents`, and you're iterating over it here.